### PR TITLE
add e2e node test for Pod hostAliases feature

### PR DESCRIPTION
**Pull request by evie404**
**What this PR does / why we need it**: adds node e2e test for #45148 

tests requested in https://github.com/kubernetes/kubernetes/issues/43632#issuecomment-298434125

**Release note**:
```release-note
NONE
```

@yujuhong @thockin

### CODE REVIEW

File: test/e2e_node/kubelet_test.go [hunk 1]
```diff
@@ -116,6 +116,62 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+			Eventually(func() string {
+				pod, err := podClient.Get(podName, metav1.GetOptions{})
+				Expect(err).To(BeNil(), fmt.Sprintf("Error getting Pod %v", err))
+				podIP := pod.Status.PodIP
+
+				expectedHostsFileContent = fmt.Sprintf(expectedHostsFileContent, podIP, podName)
+
+				rc, err := podClient.GetLogs(podName, &v1.PodLogOptions{}).Stream()
+				if err != nil {
+					return ""
```
Comment (kargakis) [inline test/e2e_node/kubelet_test.go]: Print this error?

Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: every other example seem to return blank string when there's an error from getting logs. I'm not sure what's the rationale, but just cargo-culting.

Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: @vishh could you help us understand why in other test cases it's acceptable to return blank string? sorry i'm not very familiar with the test framework used for the node e2e tests.

Comment (kargakis) [inline test/e2e_node/kubelet_test.go]: It's fine to return the empty string, logging the error is orthogonal.

Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: fixed

File: test/e2e_node/kubelet_test.go [hunk 2]
```diff
@@ -116,6 +116,62 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+							IP:        "123.45.67.89",
+							Hostnames: []string{"foo", "foo.bar"},
+						},
+					},
+				},
+			})
+
+			Eventually(func() string {
+				pod, err := podClient.Get(podName, metav1.GetOptions{})
+				Expect(err).To(BeNil(), fmt.Sprintf("Error getting Pod %v", err))
```
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: fixed

**Comment by rickypai** [comment]
issues have been addressed. if you can take a look at this PR again, that'd be great. @kargakis @mtaufen

File: test/e2e_node/kubelet_test.go [hunk 3]
```diff
@@ -116,6 +117,64 @@ var _ = framework.KubeDescribe("Kubelet", func() {
 			Expect(err).To(BeNil(), fmt.Sprintf("Error deleting Pod %v", err))
 		})
 	})
+	Context("when scheduling a busybox Pod with hostAliases", func() {
+		podName := "busybox-host-aliases" + string(uuid.NewUUID())
+		expectedHostsFileContent := `# Kubernetes-managed hosts file.
+127.0.0.1	localhost
```
Comment (yujuhong) [inline test/e2e_node/kubelet_test.go]: Do we really need to check the entire file? It should be sufficient if `/etc/hosts` contains the two lines we care about.

File: test/e2e_node/kubelet_test.go [hunk 4]
```diff
@@ -116,6 +117,64 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+			})
+
+			Eventually(func() string {
+				pod, err := podClient.Get(podName, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				podIP := pod.Status.PodIP
+
+				expectedHostsFileContent = fmt.Sprintf(expectedHostsFileContent, podIP, podName)
+
+				rc, err := podClient.GetLogs(podName, &v1.PodLogOptions{}).Stream()
```
Comment (yujuhong) [inline test/e2e_node/kubelet_test.go]: I know you copied this from other tests, but consider using `framework.GetPodLogs()` which returns a string directly. This saves you the need of reading the output yourself.

File: test/e2e_node/kubelet_test.go [hunk 5]
```diff
@@ -116,6 +117,64 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+					HostAliases: []v1.HostAlias{
+						{
+							IP:        "123.45.67.89",
+							Hostnames: []string{"foo", "foo.bar"},
+						},
+					},
+				},
+			})
+
+			Eventually(func() string {
```
Comment (yujuhong) [inline test/e2e_node/kubelet_test.go]: I'd suggest making the function returns an error, and check whether the error is nil. This way, you can generate the error you want, without thinking what should be logged (e.g., failed fetching pod logs) and whether people would notice that.

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

File: test/e2e_node/kubelet_test.go (hunk 3 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: there's currently no e2e test checking any other content of the hosts file. so i figure i can help by checking the whole thing.

fixed to just check hostAlias-related entries

File: test/e2e_node/kubelet_test.go (hunk 4 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: i haven't really figured out how to use `framework.GetPodLogs()`. keeping the old way for now.

File: test/e2e_node/kubelet_test.go (hunk 5 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: fixed to make function return error

File: test/e2e_node/kubelet_test.go (hunk 4 above)
Comment (yujuhong) [inline test/e2e_node/kubelet_test.go]: Here's an example if you'd like to give it a try:
https://github.com/kubernetes/kubernetes/blob/v1.8.0-alpha.0/test/e2e_node/security_context_test.go#L87

**Comment by rickypai** [comment]
thanks @yujuhong. should I backport this to release-1.7 branch? I think that'd be a good idea.

File: test/e2e_node/kubelet_test.go [hunk 6]
```diff
@@ -116,6 +117,51 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+					Name: podName,
+				},
+				Spec: v1.PodSpec{
+					// Don't restart the Pod since it is expected to exit
+					RestartPolicy: v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						{
+							Image:   "gcr.io/google_containers/busybox:1.24",
+							Name:    podName,
+							Command: []string{"/bin/sh", "-c", "cat /etc/hosts; sleep 240"},
```
Comment (timstclair) [inline test/e2e_node/kubelet_test.go]: nit: add a longer sleep timer and just let the namespace cleanup the pod (lots of tests use 6000 (10 minutes))

File: test/e2e_node/kubelet_test.go [hunk 7]
```diff
@@ -116,6 +117,51 @@ var _ = framework.KubeDescribe("Kubelet", func() {
...
+			})
+
+			Eventually(func() error {
+				rc, err := podClient.GetLogs(podName, &v1.PodLogOptions{}).Stream()
+				defer rc.Close()
+				if err != nil {
+					return err
+				}
+				buf := new(bytes.Buffer)
+				buf.ReadFrom(rc)
```
Comment (timstclair) [inline test/e2e_node/kubelet_test.go]: If you're going to read the whole stream, just use `\.DoRaw()` rather than `Stream()`

**Comment by timstclair** [comment]
I think 1.7 is still tracking master, so no cherrypick should be necessary.

File: test/e2e_node/kubelet_test.go (hunk 7 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: fixed

File: test/e2e_node/kubelet_test.go (hunk 6 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: fixed

**Comment by rickypai** [comment]
@k8s-bot pull-kubernetes-node-e2e test this

**Comment by rickypai** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

File: test/e2e_node/kubelet_test.go (hunk 7 above)
Comment (rickypai) [inline test/e2e_node/kubelet_test.go]: I reverted this. this was introducing panics in the tests and the logs are quite difficult to understand: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46385/pull-kubernetes-node-e2e/33666/nodelog?junit=junit_gci_05.xml&wrap=on

also, `Stream()` is used in other kubelet e2e tests, so I think it's safer to follow the pattern: https://github.com/kubernetes/kubernetes/blob/05bff300f3ed8a0bc51998ace11dad50782c84c5/test/e2e_node/kubelet_test.go#L143

**Comment by rickypai** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**